<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [400]:
import pandas as pd
import psycopg2

In [401]:
#Удаленные данные для поделючения

In [402]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [403]:
# текст запроса
query_3_1 = f'''
    SELECT
        count(id) as "Количество вакансий"
    FROM public.vacancies
'''

In [404]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\3914452570.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий    49197
Name: 0, dtype: int64

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [405]:
# текст запроса
query_3_2 = f'''
    SELECT
        count(*) as "Количество работодателей"
    FROM public.employers
'''

In [406]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\1606364727.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


Количество работодателей    23501
Name: 0, dtype: int64

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [407]:
# текст запроса
query_3_3 = f'''
    SELECT
        count(*) as "Количество регионов"
    FROM public.areas
'''

In [408]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\879057931.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


Количество регионов    1362
Name: 0, dtype: int64

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [409]:
# текст запроса
query_3_4 = f'''
    SELECT
        count(*) as "Количество сфер деятельности"
    FROM public.industries
'''

In [410]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\1874342592.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


Количество сфер деятельности    294
Name: 0, dtype: int64

***

Выводы по предварительному анализу данных  

1. Количество вакансий: 49197  

2. Количество работодателей: 23501, т.е. на каждого работодателя в среднем приходится 2,09 вакансий.

3. Количество регионов: 1362  

4. Количество сфер деятельности: 294  

Дополнительные выводы по указанным количественным показателям сделать невозможно без динамических данных (за период).

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area). Отсортируйте по количеству вакансий в порядке убывания.

In [411]:
# текст запроса
query_4_1 = f'''
    SELECT
        a.name as "Регион",
        count(v.id) as "Количество вакансий"
    FROM public.areas a
    LEFT JOIN public.vacancies v ON a.id = v.area_id
    GROUP BY a.name
    ORDER BY 2 DESC
    LIMIT 5
'''

In [412]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


Регион  Количество вакансий
0           Москва                 5333
1  Санкт-Петербург                 2851
2            Минск                 2112
3      Новосибирск                 2006
4           Алматы                 1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [413]:
# текст запроса
query_4_2 = f'''
    SELECT
        count(v.id) as "Количество вакансий"
    FROM public.vacancies v
    WHERE (salary_from IS NOT NULL) OR (salary_to IS NOT NULL)
'''

In [414]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\3822845284.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Количество вакансий    24073
Name: 0, dtype: int64

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [415]:
# текст запроса
query_4_3 = f'''
    SELECT
        round(avg(salary_from)) as "Среднее значение нижней границы ЗП",
        round(avg(salary_to)) as "Среднее верхней границы ЗП"
    FROM public.vacancies
'''

In [416]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\3930481643.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


Среднее значение нижней границы З     71065.0
Среднее верхней границы ЗП           110537.0
Name: 0, dtype: float64

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [417]:
# текст запроса
query_4_4 = f'''
    SELECT
        schedule as "Тип рабочего графика",
        employment as "Тип трудоустройства",
        count(id) as "Количество вакансий"
    FROM public.vacancies v
    GROUP BY schedule, employment
    ORDER BY 3 DESC
    OFFSET 1 LIMIT 1
'''

In [418]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\2810743503.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Тип рабочего графика    Удаленная работа
Тип трудоустройства     Полная занятость
Количество вакансий                 7802
Name: 0, dtype: object

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [419]:
# текст запроса
query_4_5 = f'''
    SELECT
        experience as "Требуемый опыт работы",
        count(id) as "Количество вакансий"
    FROM public.vacancies v
    GROUP BY experience
    ORDER BY 2
'''

In [420]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Требуемый опыт работы  Количество вакансий
0           Более 6 лет                 1337
1             Нет опыта                 7197
2         От 3 до 6 лет                14511
3    От 1 года до 3 лет                26152

***

Выводы по детальному анализу вакансий  

1. Пятёрку лидеров по количеству вакансий: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Можно предположить, что количество вакансий пропорционально населению городов.  

2. У 24073 вакансий заполнена хотя бы одна граница заработной платы.  Т.е. можно предположить, что оставшеяся часть работодателей (около половины) готовы индивидуально обсуждать с соискателем уровень заработной платы в зависимости от опыта и навыков соискателя.  

3. Среднее значение нижний границы заработной платы составляет: 71065. Среднее значение верхней границы заработной платы составляет: 110537.  Т.е. в среднем вилка заработной платы составляет около 40000.

4. Наиболее востребованная комбинация типа рабочего графика и трудоустройства: Полный день - Полная занятость.  
На втором месте: Удаленная работа - Полная занятость.  Т.е. значительная часть вакансий допускает удаленную работу, что характерно для It-отрасли.

5. Наибольшее количество вакансий для соискателей с опытом работы: От 1 года до 3 лет.  
Наименьшее: Более 6 лет. Т.е. работодатели в целом предпочитают нанимать сотрудников с меньшим опытом работы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [421]:
# текст запроса
query_5_1 = f'''
    SELECT
        e.name as "Работодатель",
        count(v.id) as "Количество вакансий"
    FROM public.employers e
    LEFT JOIN public.vacancies v ON e.id = v.employer_id
    GROUP BY e.name
    ORDER BY 2 DESC
'''

In [422]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df.iloc[[0,4],]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\3486825544.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Работодатель  Количество вакансий
0         Яндекс                 1933
4  Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [423]:
# текст запроса
query_5_2 = f'''
    SELECT
        a.name as "Регион",
        count(e.id) as "Количество работодателей",
        count(v.id) as "Количество вакансий"
    FROM public.areas a
    LEFT JOIN public.employers e ON e.area = a.id
    LEFT JOIN public.vacancies v ON v.area_id = a.id
    GROUP BY a.name
    ORDER BY 3, 2 DESC
'''

In [424]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\2401898840.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


Регион                      Россия
Количество работодателей       410
Количество вакансий              0
Name: 0, dtype: object

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [425]:
# текст запроса
query_5_3 = f'''
    SELECT
        e.name as "Работодатель",
        count(DISTINCT v.area_id) as "Количество регионов вакансий"
    FROM public.employers e
    LEFT JOIN public.vacancies v ON e.id = v.employer_id
    GROUP BY e.name
    ORDER BY 2 DESC
'''

In [426]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Работодатель  Количество регионов вакансий
0                      Яндекс                           181
1                  Ростелеком                           152
2                  Спецремонт                           116
3      Поляков Денис Иванович                            88
4                   ООО ЕФИН                             71
...                       ...                           ...
23170                  СДЕЛКА                             0
23171            Alandr Group                             0
23172                     СДК                             0
23173      Lemon Land Lombard                             0
23174                  Махина                             0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [427]:
# текст запроса
query_5_4 = f'''
    (SELECT
      id as "Id работодателя"
    FROM public.employers
    EXCEPT
    SELECT
        employer_id    
    FROM public.employers_industries)
'''

In [428]:
# текст запроса
query_5_4 = f'''
    SELECT
      count(id) as "Количество работодателей"
    FROM public.employers
    WHERE id NOT IN
        (SELECT
            employer_id    
        FROM public.employers_industries)
'''

In [429]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\4253387685.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


Количество работодателей    8419
Name: 0, dtype: int64

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [430]:
# текст запроса
query_5_5 = f'''
    SELECT
        e.name as "Работодатель",
        count(ei.industry_id) as "Количество сфер дейятельности"
    FROM public.employers e
    LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
    GROUP BY e.name
    HAVING count(ei.industry_id) = 4
    ORDER BY 1
    OFFSET 2 LIMIT 1
'''

In [431]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Работодатель  Количество сфер дейятельности
0         2ГИС                              4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [432]:
# текст запроса
query_5_6 = f'''
    SELECT
        count(distinct ei.employer_id) as "Количество работодателей"
    FROM public.employers_industries ei
    LEFT JOIN public.industries i ON i.id = ei.industry_id
    WHERE i.name = 'Разработка программного обеспечения'
'''

In [433]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\1686926634.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


Количество работодателей    3553
Name: 0, dtype: int64

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [434]:
# код для получения списка городов-милионников
cities = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')
cities = tuple(cities[1]['Город'].replace('Воронеж[a]', 'Воронеж'))
cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [435]:
# текст запроса
query_5_7 = f'''
    (SELECT
        a.name as "Регион",
        count(v.id) as "Количество вакансий"
    FROM public.areas a
    LEFT JOIN public.vacancies v ON v.area_id = a.id
    LEFT JOIN public.employers e ON e.id = v.employer_id
    WHERE a.name in {cities} AND e.name = 'Яндекс'
    GROUP BY a.name)
    UNION
    (SELECT
        'Total' as "Регион",
        count(v.id) as "Количество вакансий"
    FROM public.areas a
    LEFT JOIN public.vacancies v ON v.area_id = a.id
    LEFT JOIN public.employers e ON e.id = v.employer_id
    WHERE a.name in {cities} AND e.name = 'Яндекс')
    ORDER BY 2
'''

In [436]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


Регион  Количество вакансий
0              Омск                   21
1         Челябинск                   22
2        Красноярск                   23
3         Волгоград                   24
4             Пермь                   25
5            Казань                   25
6    Ростов-на-Дону                   25
7               Уфа                   26
8            Самара                   26
9         Краснодар                   30
10          Воронеж                   32
11      Новосибирск                   35
12  Нижний Новгород                   36
13     Екатеринбург                   39
14  Санкт-Петербург                   42
15           Москва                   54
16            Total                  485

***

Выводы по анализу работодателей  

1. На первом месте по количеству вакансий находится Яндекс, на пятом - Газпром нефть. В целом среди лидеров крупные компании из сферы телеком, банкинга и нефтегазового сетора, что говорит о наибольшей востребованности IT-услуг в указанных сферах.

2. Среди регионов, в которых нет вакансий, с наибольшим колчеством работодателей, первое место занимает Россия. Можно предположить, что причиной этого является то, что вакансия привязана (расположена) в конкретном субъекте России.

3. На первом месте среди количества регионов, в которых представлены вакансии, находится Яндекс (181 регион). Причиной этого может являться международный характер компании и/или вакансии с удаленной работой.  

4. Количество работодателей, у которых не указана сфера деятельности - 8419. Это может свидетельствовать либо высокой дифференциации сфер компаний из списка, либо об уровне культуры кадрового дела в компании, что может негативно сказаться на количестве соискателей на данную вакансию.

5. На третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности находится 2ГИС. Это говорит об различных сферах деятельности бизнеса с использованием IT-услуг специалистов, в частности для 2ГИС - исследования, разрботка, реклама, продажи.  

6. В качестве сферы деятельности "Разработка программного обеспечения указана" у 3553 компаний из 23501. Т.е. около 15% работодателей разрабатывают собственные сервисы, приложения и продукты.

7. Компания Яндекс ищет соискателей во всех 16 городах-миллиониках. При этом можно проследить зависимость между количеством населения города и количеством вакансий. Общее количество вакансий Яндекса, приходящиеся на города миллионники 485 из 1933 всех вакансий компании, т.е. около 25%.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [437]:
# текст запроса
query_6_1 = f'''
    SELECT
        count(v.id) as "Количество вакансий"
    FROM public.vacancies v
    /* Ищем без учета регистра*/
    WHERE v.name ILIKE '%data%' or v.name ILIKE '%данн%'
'''

In [438]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\131442351.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


Количество вакансий    1771
Name: 0, dtype: int64

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [439]:
# текст запроса
query_6_2 = f'''
    SELECT
        count(v.id) as "Количество вакансий"
    FROM public.vacancies v
    /* Ищем без учета регистра*/
    where (v.name ilike '%data scientist%' or
        v.name ilike '%data science%' or
        v.name ilike '%исследователь данных%' or
        v.name ilike '%machine learning%' or
        v.name ilike '%машинн%обучен%' or
        (v.name not like '%HTML%' and v.name like '%ML%')) and
        (v.name ilike '%junior%' or
        v.experience ilike '%Нет опыта%' or
        v.employment ilike '%Стажировка%')
'''

In [440]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\1247029863.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


Количество вакансий    51
Name: 0, dtype: int64

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [441]:
# текст запроса
query_6_3 = f'''
    SELECT
        count(v.id) as "Количество вакансий"
    FROM public.vacancies v
    /* Ищем без учета регистра*/
    where (v.name ilike '%data scientist%' or
        v.name ilike '%data science%' or
        v.name ilike '%исследователь данных%' or
        v.name ilike '%machine learning%' or
        v.name ilike '%машинн%обучен%' or
        (v.name not like '%HTML%' and v.name like '%ML%')) and
        (v.key_skills like '%SQL%' or
        v.key_skills ilike '%postgres%')
'''

In [442]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\1957183952.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


Количество вакансий    201
Name: 0, dtype: int64

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [443]:
# текст запроса
query_6_4 = f'''
    SELECT
        count(v.id) as "Количество вакансий"
    FROM public.vacancies v
    /* Ищем без учета регистра*/
    where (v.name ilike '%data scientist%' or
        v.name ilike '%data science%' or
        v.name ilike '%исследователь данных%' or
        v.name ilike '%machine learning%' or
        v.name ilike '%машинн%обучен%' or
        (v.name not like '%HTML%' and v.name like '%ML%')) and
        (v.key_skills ilike '%python%')
'''

In [444]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\2997710971.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


Количество вакансий    351
Name: 0, dtype: int64

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [445]:
query_6_5 = f'''
    SELECT
        /*
        Заменяем разделители с помощью REPLACE()
        Считаем количество слов какразность между количеством сиволов с разделителями и без
        Вычесляем среднее и округляем
        */
        round(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) as "Среднее количество ключевых навыков"
    FROM public.vacancies
    where key_skills is not null and (name ilike '%data scientist%' or
        name ilike '%data science%' or
        name ilike '%исследователь данных%' or
        name ilike '%machine learning%' or
        name ilike '%машинн%обучен%' or
        (name not like '%HTML%' and name like '%ML%'))
'''

In [446]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df.iloc[0]

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\3476253169.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Среднее количество ключевых навык    6.41
Name: 0, dtype: float64

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [447]:
# текст запроса
query_6_6 = f'''
    SELECT
        distinct experience as "Тип опыта",
        /* Считаем среднюю ЗП используя coalesce() для замены null на 0 */
        round(AVG(coalesce((salary_from+salary_to)/2, salary_from, salary_to)),0) as "Средняя зарплата"
    FROM public.vacancies
    where (salary_from is not null or salary_to is not null) and
        (name ilike '%data scientist%' or
        name ilike '%data science%' or
        name ilike '%исследователь данных%' or
        name ilike '%machine learning%' or
        name ilike '%машинн%обучен%' or
        (name not like '%HTML%' and name like '%ML%'))
        GROUP BY 1
'''

In [448]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


Тип опыта  Средняя зарплата
0           Нет опыта           74643.0
1  От 1 года до 3 лет          139675.0
2       От 3 до 6 лет          243115.0

***

Выводы по предметному анализу

1. Вакансий, имеющих отношения к данным 1771 из 49197, т.е. около 4% вакансий включают требований к обработке данных. Можно сделать вывод, что специалисты в области работы с данными не являются восстребованными или работа с данными не выделяется как ключевая задачи в вакансиях. Можно предположить, что это связано с молодостью направления data science. В целом, желательно, посмотреть количество вакансий связанных с data science в динамике (т.е. с учетом даты размещения).

2. Для начинающего дата-сайентиста общеее количество вакансий 51, это около 3% от всех вакансий связанных с работой с данными и 0,1% от всех вакансий в IT-сфере. Т.е. в целом можно сделать вывод о невысокой востребованности дата-сайентистов с низким опытом на рынке труда в период времени охваченной данной выборкой.

3. Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres 201 или 11% от всех DS вакансий.

4. Количество вакансий, в которых в качестве ключевого навыка указан Python 351 или 20%, что свидетельствует о большей популярности/применимости Python в DS по сравнению с SQL и postgres.

5. Количество ключевых навыков в среднем указывают в вакансиях для DS 6.41. Тут интересно было бы провести анализ навыков по популярности.

6. В среднем дата-сайентист с опытом работы от 3 до 6 лет может рассчитывать на зарплату в 243115, что значительно выше (больше, чем в двое) средней зарплта по верхней границе в IT-сфере (110537). Кроме того, если посмотреть на зависимость зарпалаты от опыта, то можно сделать вывод, что при переходе от каждого типа опыта (в целом можно говорить о грейде junior/middle/senior) зарплата в среднем растет на 82%.

# Общий вывод по проекту

In [449]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Итоги исследования    

Была проанализирована выборка из 49197 вакансий от 23501. В целом распределение вакансий по регионам пропорционально численности населения в регионах. А самым активным рынком треда в IT-секторе является Москва. Самым крупным работодателем по количеству ваканчий является Сбер, в целом первую пятерку рейтинга работодателей по количеству ваканчий занимают компании, ведущие свою делятельность в секторах телеком, банкинга и ТЭК. При этом доля вакансий, связанных с DS состовляет 4% от общего числа вакансий.

Было бы интересно, провести дополнительные исследования в динамике, например проанализировать изменение числа вакансий DS во времени. Однако, к сожалению данных о дате размещения вакансии в выборке не представлено.  

На основании имеющихся данных предлагается провести следующие дополнительные исследования:  
    Выяснить в какой сфере деятельности задействовано наибольшее количество компаний    
    Выяснить в какой сфере деятельности задействовано наибольшее количество вакансий    
    Провести более детальный анализ самого активного ранка труда - Москвы:   
        Выяснить среднюю зарплату по региону;   
        Определить рейтинг работодателй по количеству вакансий в регионе;    
        Выяснить тип рабочего графика, наиболее популярного в регионе.    

## 7. Дополнительные исследования

7.1. Выяснить в какой сфере деятельности задействовано наибольшее количество компаний 

In [450]:
# текст запроса
query_7_1 = f'''
    SELECT
        i.name as "Сфера деятельности деятельности",
        count(ei.employer_id) as "Количество работодателей"
    FROM public.industries i
    LEFT JOIN public.employers_industries ei ON i.id = ei.industry_id
    GROUP BY 1
    ORDER BY 2 DESC
'''

In [451]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\2931715827.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


Сфера деятельности деятельности  \
0                  Разработка программного обеспечения   
1    Системная интеграция,  автоматизации технологи...   
2    Интернет-компания (поисковики, платежные систе...   
3    Маркетинговые, рекламные, BTL, дизайнерские, E...   
4                                Консалтинговые услуги   
..                                                 ...   
289        Лифтовое хозяйство (монтаж, сервис, ремонт)   
290                                  Ритуальные услуги   
291              Ботанический сад, зоопарк, заповедник   
292                                          Автошкола   
293             Ассоциация в сфере культуры, искусства   

     Количество работодателей  
0                        3553  
1                        2993  
2                        1675  
3                         798  
4                         662  
..                        ...  
289                         4  
290                         4  
291                         3  
292                         3  
293                         1  

[294 rows x 2 columns]

Вывод  

В целом можно сделать вывод, что наибольшее количество компаний заняты в сферах разработки программного обеспечения и систмной интеграции.

7.2. Выяснить в какой сфере деятельности задействовано наибольшее количество вакансий

In [452]:
# текст запроса
query_7_2 = f'''
    SELECT
        i.name as "Сфера деятельности деятельности",
        count(distinct v.employer_id) as "Количество вакансий"
    FROM public.industries i
    LEFT JOIN public.employers_industries ei ON i.id = ei.industry_id
    LEFT JOIN public.vacancies v ON ei.employer_id = v.employer_id
    GROUP BY 1
    ORDER BY 2 DESC
'''

In [453]:
# результат запроса
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\746169716.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


Сфера деятельности деятельности  Количество вакансий
0                  Разработка программного обеспечения                 2592
1    Системная интеграция,  автоматизации технологи...                 2160
2    Интернет-компания (поисковики, платежные систе...                 1157
3                                Консалтинговые услуги                  435
4    Маркетинговые, рекламные, BTL, дизайнерские, E...                  399
..                                                 ...                  ...
289                                          Химчистка                    3
290             Ассоциация в сфере культуры, искусства                    1
291                                  Ритуальные услуги                    1
292                                          Автошкола                    1
293              Ботанический сад, зоопарк, заповедник                    0

[294 rows x 2 columns]

Вывод

На основании полученной выборке получаем, что результаты по количеству вакансий в различный сферах в целом аналогичны/коррелируют с результатами по количеству работодателй.

### 7.3. Провести более детальный анализ самого активного ранка труда - Москвы.

7.3.1. Выяснить среднюю зарплату по региону Москва и сравнии его с средним значением ЗП по всем вакансиям.

In [454]:
# текст запроса
query_7_3_1 = f'''
    (SELECT
        (round(avg(v.salary_from)) + round(avg(v.salary_to)))/2 as "Среднее значение ЗП",
        'Москва' as "Выборка"
    FROM public.vacancies v
    LEFT JOIN public.areas a ON a.id = v.area_id
    WHERE a.name = 'Москва')
    UNION ALL
    (SELECT
        (round(avg(v.salary_from)) + round(avg(v.salary_to)))/2 as "Среднее значение ЗП",
        'Все вакансии' as "Выборка"
    FROM public.vacancies v)
'''

In [455]:
# результат запроса
df = pd.read_sql_query(query_7_3_1, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\240309950.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3_1, connection)


Среднее значение ЗП       Выборка
0             137001.5        Москва
1              90801.0  Все вакансии

Вывод

Средняя зарплата выше в Москве (137001.5) чем в среднем по всем регионам (90801.0) на 46200 или на 50%.

7.3.2. Определить рейтинг работодателй по количеству вакансий в регионе.

In [456]:
# текст запроса
query_7_3_2 = f'''
    SELECT
        e.name as "Работодатель",
        count(v.id) as "Количество вакансий"
    FROM public.employers e
    LEFT JOIN public.vacancies v ON e.id = v.employer_id
    LEFT JOIN public.areas a ON a.id = v.area_id
    WHERE a.name = 'Москва'
    GROUP BY e.name
    ORDER BY 2 DESC
'''

In [457]:
# результат запроса
df = pd.read_sql_query(query_7_3_2, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\1951601484.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3_2, connection)


Работодатель  Количество вакансий
0                    СБЕР                  205
1                     МТС                  122
2                      VK                   79
3                    Ozon                   72
4                  Яндекс                   54
...                   ...                  ...
2315  Инаномо Технолоджис                    1
2316               Инвайт                    1
2317             ИндаСофт                    1
2318            Инжсиском                    1
2319  Инноватор Косметикс                    1

[2320 rows x 2 columns]

Вывод

На первом местре по количеству вакансий в регионе Москва находится Сбер. Рейтинг работодателей по количеству вакансий в Москве значительно отличается от рынка в целом.

7.3.3. Выяснить тип рабочего графика, наиболее популярного в регионе.

In [458]:
# текст запроса
query_7_3_3 = f'''
    SELECT
        v.schedule as "Тип рабочего графика",
        v.employment as "Тип трудоустройства",
        count(v.id) as "Количество вакансий"
    FROM public.vacancies v
    LEFT JOIN public.areas a ON a.id = v.area_id
    WHERE a.name = 'Москва'
    GROUP BY schedule, employment
    ORDER BY 3 DESC
'''

In [459]:
# результат запроса
df = pd.read_sql_query(query_7_3_3, connection)
df

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_14996\1065159463.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3_3, connection)


Тип рабочего графика  Тип трудоустройства  Количество вакансий
0           Полный день     Полная занятость                 3968
1      Удаленная работа     Полная занятость                  941
2         Гибкий график     Полная занятость                  178
3           Полный день           Стажировка                   85
4      Удаленная работа  Частичная занятость                   33
5        Сменный график     Полная занятость                   30
6         Гибкий график           Стажировка                   27
7           Полный день  Частичная занятость                   21
8      Удаленная работа     Проектная работа                   12
9         Гибкий график  Частичная занятость                   12
10     Удаленная работа           Стажировка                   12
11          Полный день     Проектная работа                   10
12        Гибкий график     Проектная работа                    2
13       Вахтовый метод     Полная занятость                    1
14       Сменный график           Стажировка                    1

Вывод    

Требования к типу трудоустройства и типу рабочего графика в соотношении с количеством вакансий не отличаются от рынка в целом.